### PR TITLE
feat: add coordinator populator, add base populator

### LINKED ISSUE #407: Create populators for coordinators

### PR DESCRIPTION
**Pull request by Auch-Auch**
Closes #407

### CODE REVIEW

File: django_webapps/scripts/db_population/populate_coordinators.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    "name": 1,
+    "email": 2,
+    "tg": 3,
+    "timezone": 4,
+}
+
+
+@dataclass
+class CoordinatorData(BasePersonEntityData):
+    project_status: CoordinatorProjectStatus = CoordinatorProjectStatus.WORKING_OK
```
Comment (Auch-Auch) [inline django_webapps/scripts/db_population/populate_coordinators.py:44]: Maybe introduce some Legacy status?

File: django_webapps/scripts/db_population/populate_coordinators.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+)
+from django_webapps.scripts.db_population.parsers import common_parsers  # noqa: E402
+from django_webapps.scripts.db_population.utils import (  # noqa: E402
+    get_args,
+    get_logger,
+    load_csv_data,
+)
+
+logger = get_logger("coordinators.log")
+LIST_OF_ADMINS_IDS = [22, 42]
```
Comment (Auch-Auch) [inline django_webapps/scripts/db_population/populate_coordinators.py:29]: Or add new column to the google sheet

File: django_webapps/scripts/db_population/base_populator.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    utc_timedelta: datetime.timedelta = datetime.timedelta(hours=0)
+
+    # use default values
+    def __setattr__(self, name: str, value: Any) -> None:
+        if getattr(self, name, None) is not None and value is None:
+            return
+        super().__setattr__(name, value)
+
+
+class BasePersonCsvPopulator(ABC):
```
Comment (Auch-Auch) [inline django_webapps/scripts/db_population/base_populator.py:37]: Have not fought about it too much, just wanted to put common method somewhere

File: django_webapps/scripts/db_population/populate_coordinators.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    "tg": 3,
+    "timezone": 4,
+}
+
+
+@dataclass
+class CoordinatorData(BasePersonEntityData):
+    project_status: CoordinatorProjectStatus = CoordinatorProjectStatus.WORKING_OK
+    last_name: str = ""
+    is_validated: bool = False
```
Comment (Auch-Auch) [inline django_webapps/scripts/db_population/populate_coordinators.py:46]: New column?

File: django_webapps/scripts/db_population/base_populator.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        self.csv_data = self._pre_process_data(csv_data[1:])
+        self._logger = logger
+        self._dry = dry
+        self._current_entity = None
+
+    def run(self) -> None:
+        for entity in tqdm(self.csv_data, desc=f"Processing {self.entity_name}s..."):
+            tid = entity[self._column_to_id["tid"]]
+            self._logger.info("======================================================= ")
+            self._logger.info(f"Parsing {self.entity_name} with tid {tid}")
```
[the remark this answers is not in the document]
Comment (Auch-Auch) [reply django_webapps/scripts/db_population/base_populator.py:58]: It only a name for the id column, the real value is under an index but okay let's rename it

File: django_webapps/scripts/db_population/populate_coordinators.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    status_since: datetime.datetime = timezone.now()
+
+
+class CoordinatorPopulator(BasePersonCsvPopulator):
+    entity_name: str = "coordinator"
+
+    def _pre_process_data(self, csv_data: CsvData) -> CsvData:
+        self.header[1] = "name"
+        self.header[4] = "timezone"
+        csv_data = csv_data[2:]
```
Comment (lemontree210) [inline django_webapps/scripts/db_population/populate_coordinators.py:58]: just checking: is it consciously set to `2:`, not `1:`?

File: django_webapps/scripts/db_population/utils.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    file_handler = logging.FileHandler(file_name)
+    file_handler.setFormatter(logging.Formatter("%(levelname)s - %(message)s"))
+    logger.addHandler(file_handler)
+    return logger
+
+
+def get_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--input_csv", "-i", type=str, required=True, help="Input csv file with teachers"
```
Comment (lemontree210) [inline django_webapps/scripts/db_population/utils.py:18]:
```suggestion
        "--input_csv", "-i", type=str, required=True, help="Input csv file with coordinators, groups, students or teachers"
```

**Review by lemontree210 — COMMENTED** [review]
This is the rest of my review :)

File: django_webapps/scripts/db_population/populate_coordinators.py (hunk 1 above)
Comment (Auch-Auch) [reply django_webapps/scripts/db_population/populate_coordinators.py:44]: There is not such status

File: django_webapps/scripts/db_population/populate_coordinators.py (hunk 6 above)
Comment (Auch-Auch) [reply django_webapps/scripts/db_population/populate_coordinators.py:58]: It is correct since the first two columns are some trash data

Comment (Auch-Auch) [reply django_webapps/scripts/db_population/populate_coordinators.py:58]: It is based on the csv from google sheet

Comment (Auch-Auch) [reply django_webapps/scripts/db_population/populate_coordinators.py:58]: maybe a better approach would be to fix the google sheet, not sure

File: django_webapps/scripts/db_population/populate_coordinators.py (hunk 1 above)
Comment (lemontree210) [reply django_webapps/scripts/db_population/populate_coordinators.py:42]: No, I mean that if there are some legacy statuses in Google Sheets, we can put a comment saying `"Status before conversion": [some status from Google Sheet]`

Comment (Auch-Auch) [reply django_webapps/scripts/db_population/populate_coordinators.py:42]: Yes, I meant there is no any status in the google sheet

Comment (lemontree210) [reply django_webapps/scripts/db_population/populate_coordinators.py:42]: oh, sorry. So by "introducing legacy status" you meant that they should go to Google Sheets and add something there? OK, got it. Well, let's just put `WORKING_OK` for now. After that, during conversion of groups, we should be able to change them to more precise ones where necessary.

File: django_webapps/scripts/db_population/populate_coordinators.py (hunk 6 above)
Comment (lemontree210) [reply django_webapps/scripts/db_population/populate_coordinators.py:57]: nah :)

File: django_webapps/scripts/db_population/base_populator.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+                utc_timedelta=entity_data.utc_timedelta,
+            )
+        except IntegrityError as _:
+            self._logger.warning(f"{self.entity_name} might be a duplicate")
+        return None
+
+    def _create_language_and_levels(
+        self, levels: list[str]
+    ) -> QuerySetAny[LanguageAndLevel, LanguageAndLevel]:
+        if len(levels) == 0:
```
Comment (Roxe322) [inline django_webapps/scripts/db_population/base_populator.py:132]: Я бы так написал:

```
qs = LanguageAndLevel.objects.filter(language__name="English")
if levels:
    qs = qs.filter(level_id__in=levels)
return qs
```

File: django_webapps/scripts/db_population/base_populator.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+            return LanguageAndLevel.objects.filter(language__name="English")
+        return LanguageAndLevel.objects.filter(language__name="English", level__id__in=levels)
+
+    def _create_availability_slots(
+        self, weekly_slots: list[list[tuple[int, int]]]
+    ) -> list[DayAndTimeSlot]:
+        result: list[DayAndTimeSlot] = []
+        for day_index, slots in enumerate(weekly_slots):
+            for slot in slots:
```
Comment (Roxe322) [inline django_webapps/scripts/db_population/base_populator.py:141]: `for from_hour, to_hour in slots:`
Тогда ниже можно заменить slot[0] и slot[1]

File: django_webapps/scripts/db_population/parsers/common_parsers.py [hunk 10]
```diff
@@ -87,3 +87,11 @@ def parse_telegram_name(tg_str: str) -> str | None:
         return f"@{match_result.group(0)}"
 
     raise ValueError(f"tg: {tg_str} is not valid")
+
+
+def parse_name(name_str: str) -> str:
+    min_length = 2
```
Comment (Roxe322) [inline django_webapps/scripts/db_population/parsers/common_parsers.py:93]: Вынеси в константы?

File: django_webapps/scripts/db_population/populate_coordinators.py (hunk 2 above)
Comment (lemontree210) [reply django_webapps/scripts/db_population/populate_coordinators.py:29]:
```suggestion
LIST_OF_ADMINS_IDS = [1, 22, 39, 42, 225, 238, 208]
```

**Review by lemontree210 — APPROVED** [review]
LGTM

**Pull request merged**